src/main/kotlin/com/amazon/opendistroforelasticsearch/indexstatemanagement/step/snapshot/AttemptSnapshotStep.kt

```diff
@@ -0,0 +1,115 @@
+import org.elasticsearch.action.admin.cluster.snapshots.create.CreateSnapshotResponse
+import org.elasticsearch.client.Client
+import org.elasticsearch.cluster.service.ClusterService
+import org.elasticsearch.rest.RestStatus
+import org.elasticsearch.snapshots.ConcurrentSnapshotExecutionException
+import java.time.LocalDateTime
+import java.time.ZoneId
+import java.time.format.DateTimeFormatter
+import java.util.*
+
+class AttemptSnapshotStep(
+    val clusterService: ClusterService,
+    val client: Client,
+    val config: SnapshotActionConfig,
+    managedIndexMetaData: ManagedIndexMetaData
+) : Step(name, managedIndexMetaData) {
+
+    private val logger = LogManager.getLogger(javaClass)
+    private var stepStatus = StepStatus.STARTING
+    private var info: Map<String, Any>? = null
+
+    override fun isIdempotent() = true
+
+    @Suppress("TooGenericExceptionCaught")
+    override suspend fun execute() {
+        try {
+            logger.info("Executing snapshot on ${managedIndexMetaData.index}")
+            val snapshotName = config
+                    .snapshot
+                    .plus("-")
+                    .plus(LocalDateTime
+                            .now(ZoneId.of("UTC"))
+                            .format(DateTimeFormatter.ofPattern("uuuu.MM.dd-HH:mm:ss.SSS", Locale.ROOT)))
+            val mutableInfo = mutableMapOf("snapshotName" to snapshotName)
+
+            val createSnapshotRequest = CreateSnapshotRequest()
+                    .userMetadata(mapOf("snapshot_created" to "Open Distro for Elasticsearch Index Management"))
+                    .indices(managedIndexMetaData.index)
+                    .snapshot(snapshotName)
+                    .repository(config.repository)
+                    .waitForCompletion(false)
+
+            if (config.includeGlobalState != null) {
+                createSnapshotRequest.includeGlobalState(config.includeGlobalState)
+            }
+
+            val response: CreateSnapshotResponse = client.admin().cluster().suspendUntil { createSnapshot(createSnapshotRequest, it) }
+            when (response.status()) {
+                RestStatus.ACCEPTED -> {
+                    stepStatus = StepStatus.COMPLETED
+                    mutableInfo["message"] = "Snapshot creation started and is still in progress for index: ${managedIndexMetaData.index}"
+                }
+                RestStatus.OK -> {
+                    stepStatus = StepStatus.COMPLETED
+                    mutableInfo["message"] = "Snapshot created for index: ${managedIndexMetaData.index}"
+                }
+                else -> {
+                    stepStatus = StepStatus.FAILED
+                    mutableInfo["message"] = "There was an error during snapshot creation for index: ${managedIndexMetaData.index}"
+                    mutableInfo["cause"] = response.toString()
+                }
+            }
+            info = mutableInfo.toMap()
+        } catch (e: ConcurrentSnapshotExecutionException) {
```

Comment:
Since this step is considered `idempotent` how do we know if the existing snapshot creation in progress is from this step or from user triggered snapshot?

We should consider the StepStatus to be `completed` if the snapshot is already in progress by this step and move on to `WaitForSnapshot`

Reply:
@jinsoor-amzn ConcurrentSnapshotExecutionException is thrown when any snapshot is in progress, possibly one that does not include this index.

Did not see the isIdempotent was set to `true` here, good catch.

We probably want to change it to `false`. Considering this instead causes a `new` snapshot to be created even if the previous one was already kicked off and completed I'm not sure I'd consider it idempotent.. although since these are technically just diffs built on top of each other I'm not sure it would really hurt.

For clarification  @dabr-grapeup and @JavierAbrego, we set steps to idempotent when it really doesn't matter if it gets called again if the previous execution failed to fully complete. Things like setting the index settings (readonly, readwrite) can be called multiple times with the same response. Same with close/open, these can be called multiple times with no side effects. Calling close on an already closed index simply returns a 200 ack:true. Calling snapshot multiple times can end up with multiple snapshots being created which might be OK for some people and not OK for others.